python/ray/data/tests/test_dataset_formats.py

```diff
@@ -409,6 +409,117 @@ def test_read_tfrecords(ray_start_regular_shared, tmp_path):
+    # from this dataset.
+
+    expected_records = [
+        # Record one (corresponding to row one).
+        tf.train.Example(
+            features=tf.train.Features(
+                feature={
+                    "int_item": tf.train.Feature(
+                        int64_list=tf.train.Int64List(value=[1])
+                    ),
+                    "int_list": tf.train.Feature(
+                        int64_list=tf.train.Int64List(value=[2, 2, 3])
+                    ),
+                    "float_item": tf.train.Feature(
+                        float_list=tf.train.FloatList(value=[1.0])
+                    ),
+                    "float_list": tf.train.Feature(
+                        float_list=tf.train.FloatList(value=[2.0, 3.0, 4.0])
+                    ),
+                    "bytes_item": tf.train.Feature(
+                        bytes_list=tf.train.BytesList(value=[b"abc"])
+                    ),
+                    "bytes_list": tf.train.Feature(
+                        bytes_list=tf.train.BytesList(value=[b"abc", b"1234"])
+                    ),
+                }
+            )
+        ),
+        # Record two (corresponding to row two).
+        tf.train.Example(
+            features=tf.train.Features(
+                feature={
+                    "int_item": tf.train.Feature(
+                        int64_list=tf.train.Int64List(value=[2])
+                    ),
+                    "int_list": tf.train.Feature(
+                        int64_list=tf.train.Int64List(value=[3, 3, 4])
+                    ),
+                    "float_item": tf.train.Feature(
+                        float_list=tf.train.FloatList(value=[2.0])
+                    ),
+                    "float_list": tf.train.Feature(
+                        float_list=tf.train.FloatList(value=[2.0, 2.0, 3.0])
+                    ),
+                    "bytes_item": tf.train.Feature(
+                        bytes_list=tf.train.BytesList(value=[b"def"])
+                    ),
+                    "bytes_list": tf.train.Feature(
+                        bytes_list=tf.train.BytesList(value=[b"def", b"1234"])
+                    ),
+                }
+            )
+        ),
+    ]
+
+    # Perform the test.
+    # Write the dataset to a .tfrecords file.
+    ds.write_tfrecords(tmp_path)
+
+    # Read the Examples back out from the .tfrecords file.
+    # This follows the offical TFRecords tutorial:
+    # https://www.tensorflow.org/tutorials/load_data/tfrecord#reading_a_tfrecord_file_2
+
+    filenames = sorted(os.listdir(tmp_path))
+    filepaths = [os.path.join(tmp_path, filename) for filename in filenames]
+    raw_dataset = tf.data.TFRecordDataset(filepaths)
+
+    tfrecords = []
+    for raw_record in raw_dataset:
+        example = tf.train.Example()
+        example.ParseFromString(raw_record.numpy())
+        tfrecords.append(example)
+
+    assert tfrecords == expected_records
+
+
```

Comment:
Now that we have a write-side of the TFRecords I/O layer, might be worth breaking these tests out into its own testing module, e.g. `test_dataset_tfrecords.py`.